src/plugins/discover/public/application/main/components/sidebar/discover_sidebar_responsive.tsx

```diff
@@ -111,38 +114,54 @@ export interface DiscoverSidebarResponsiveProps {
  */
 export function DiscoverSidebarResponsive(props: DiscoverSidebarResponsiveProps) {
   const services = useDiscoverServices();
+  const { data, dataViews, core } = services;
   const isPlainRecord = useAppStateSelector(
     (state) => getRawRecordType(state.query) === RecordRawType.PLAIN
   );
   const { selectedDataView, onFieldEdited, onDataViewCreated } = props;
   const [fieldFilter, setFieldFilter] = useState(getDefaultFieldFilter());
   const [isFlyoutVisible, setIsFlyoutVisible] = useState(false);
-  /**
-   * fieldCounts are used to determine which fields are actually used in the given set of documents
-   */
-  const fieldCounts = useRef<Record<string, number> | null>(null);
-  if (fieldCounts.current === null) {
-    fieldCounts.current = calcFieldCounts(props.documents$.getValue().result!, selectedDataView);
-  }
+  const [allFields, setAllFields] = useState<DataViewField[] | null>(null);
+  const [fieldCounts, setFieldCounts] = useState<Record<string, number> | null>(null);
+  const dataViewFields = selectedDataView?.fields;
 
-  const [documentState, setDocumentState] = useState(props.documents$.getValue());
   useEffect(() => {
-    const subscription = props.documents$.subscribe((next) => {
-      if (next.fetchStatus !== documentState.fetchStatus) {
-        if (next.result) {
-          fieldCounts.current = calcFieldCounts(next.result, selectedDataView!);
-        }
-        setDocumentState({ ...documentState, ...next });
+    const subscription = props.documents$.subscribe((documentState) => {
+      if (documentState?.fetchStatus === FetchStatus.COMPLETE) {
+        setFieldCounts(calcFieldCounts(documentState.result));
+      } else {
+        setFieldCounts(null);
       }
     });
     return () => subscription.unsubscribe();
-  }, [props.documents$, selectedDataView, documentState, setDocumentState]);
+  }, [props.documents$, setAllFields, setFieldCounts]);
 
   useEffect(() => {
-    // when data view changes fieldCounts needs to be cleaned up to prevent displaying
-    // fields of the previous data view
-    fieldCounts.current = {};
-  }, [selectedDataView]);
+    setAllFields(getDataViewFieldList(selectedDataView, fieldCounts, isPlainRecord));
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [dataViewFields, fieldCounts, setAllFields, isPlainRecord]);
```

Comment:
quick question, wouldn't it also work like this? Locally it seemed to be the case (L126 would be no longer necessary)

```TypeScript
  useEffect(() => {
    setAllFields(getDataViewFieldList(selectedDataView, fieldCounts, isPlainRecord));
  }, [selectedDataView, fieldCounts, setAllFields, isPlainRecord]);
```

Reply:
We can also keep it when it doubt ... more likely it's fieldCounts triggering the update?

Reply:
Most probably. 

Apparently, we trigger documents fetching after modifying a field even if `discover:searchOnPageLoad` is off. Looks like a bug in the existing functionality, no?

Reply:
I agree it looks like this

Reply:
since it's an existing bug, i think its worth an issue?